ql/ql/src/queries/bugs/RestrictiveClosure.ql

```diff
@@ -0,0 +1,53 @@
+/**
+ * @name Restrictive transitive closure
+ * @description The transitive closure operation might restrict the type even when taking zero steps.
+ * @kind problem
+ * @problem.severity warning
+ * @id ql/restrictive-transitive-closure
+ * @tags correctness
+ *       maintainability
+ * @precision high
+ */
+
+import ql
+
+Class base(Call c) {
+  result = c.(MemberCall).getBase().getType().getDeclaration()
+  or
+  not c instanceof MemberCall and
+  result = c.getTarget().(ClassPredicate).getParent()
+}
+
+Call closureCall() { result.isClosure("*") }
+
+Class return(Call c) { result = c.getTarget().getReturnType().getDeclaration() }
+
+Class superClass(Class c) { result = c.getASuperType().getResolvedType().getDeclaration() }
+
+from Call c, Class return, Class base
+where
+  c = closureCall() and
+  return = return(c) and
+  base = base(c) and
+  return != base and
+  // We aren't restricted anyway from the sorounding code.
+  not superClass*(base) = return and
```

Comment:
Am I missing something subtle here, or does this not imply `return != base` (from line 32)?